File: 0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.py
import heapq

class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
       
        def euclid_distance(x, y):
            return (x * x + y * y) ** (1/2)
        
        heap = []
        answer = []
        for x, y in points:
            heapq.heappush(heap, [euclid_distance(x, y), x, y])
        while k:
            element = heapq.heappop(heap)
            answer.append([element[1], element[2]])
            k -= 1
        return answer

File: 0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.kt
class Solution {
    data class Point(
        var distance: Long,
        var point: IntArray
    )

    fun kClosest(points: Array<IntArray>, k: Int): Array<IntArray> {
        val answer = mutableListOf<IntArray>()
        val pq = PriorityQueue<Point>(compareBy { x -> x.distance})

        for (point in points){
            pq.add(Point(point[0].toLong() * point[0] + point[1].toLong() * point[1], point))            
        }
        var count = k
        while (count > 0) {
            answer.add(pq.poll().point)
            count--
        }
        return answer.toTypedArray()
    }
}